File: src/spark-examples/sparknormal-examples/SparkOnClickHouseJavaExample/src/main/java/com/huawei/bigdata/spark/examples/SparkOnClickHouseExample.java
package com.huawei.bigdata.spark.examples;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Properties;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.SparkSession;

import com.clickhouse.jdbc.ClickHouseConnection;
import com.clickhouse.jdbc.ClickHouseDriver;
import com.clickhouse.jdbc.ClickHouseStatement;

/**
 * 功能描述 SparkOnClickHouseJavaExample
 *
 * @since 2022-03-17
 */
public class SparkOnClickHouseExample {

    private static final String DRIVER = "com.clickhouse.jdbc.ClickHouseDriver";

    public static void main(String[] args) throws SQLException {
        if (args.length < 3) {
            System.out.println("Missing parameters, need ckJdbcUrl, ckDBName, ckTableName");
            System.exit(0);
        }

        SparkSession spark = SparkSession
                .builder()
                .appName("SparkOnClickHouseJavaExample")
                .enableHiveSupport()
                .getOrCreate();

        String jdbcUrl = args[0];
        String ckDBName = args[1];
        String ckTableName = args[2];
        String user = args[3];

        Properties props = new Properties();
        props.put("driver", DRIVER);
        props.put("user", user);

        ClickHouseDriver ckDriver = new ClickHouseDriver();
        ClickHouseConnection ckConnect = ckDriver.connect(jdbcUrl, props);
        ClickHouseStatement ckStatement = ckConnect.createStatement();

        String allDBs = "show databases";
        ResultSet rs = ckStatement.executeQuery(allDBs);
        while (rs.next()) {
            String dbName = rs.getString(1);
            System.out.println("dbName: " + dbName);
        }

        Long startTime = System.currentTimeMillis();
        clickHouseExecute(ckStatement, ckDBName, ckTableName);
        System.out.println("[Elapsed]" + (System.currentTimeMillis() - startTime));

        Dataset ckData = spark.read()
                .format("jdbc")
                .option("url", jdbcUrl)
                .option("user", user)
                .option("driver", DRIVER)
                .option("dbtable", ckDBName+"."+ckTableName)
                .load();

        ckData.show();

        ckData.registerTempTable("ckTempTable");

        Dataset newCkData = spark.sql("select EventDate,cast(id+20 as decimal(20,0)) as id,name,age,address from ckTempTable");

        newCkData.show();
        newCkData.write().mode("Append").jdbc(jdbcUrl, ckDBName + "." + ckTableName, props);

        spark.stop();
    }

    private static void clickHouseExecute(ClickHouseStatement ckStatement, String clickHouseDB, String clickHouseTable)
            throws SQLException {
        String createDB = "CREATE DATABASE " + clickHouseDB + " ON CLUSTER default_cluster";
        String createTable = "CREATE TABLE " + clickHouseDB + "." + clickHouseTable + " ON CLUSTER default_cluster " +
                "(`EventDate` DateTime,`id` UInt64,`name` String,`age` UInt8,`address` String)" +
                "ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/" + clickHouseDB + "/" + clickHouseTable + "', '{replica}')" +
                "PARTITION BY toYYYYMM(EventDate) ORDER BY id";

        String insertData = "insert into " + clickHouseDB + "." + clickHouseTable + " (id, name,age,address) values (1, 'zhangsan',17,'xian'), (2, 'lisi',36,'beijing')";

        ckStatement.execute(createDB);
        ckStatement.execute(createTable);
        ckStatement.execute(insertData);
    }
}

File: src/spark-examples/sparksecurity-examples/SparkOnClickHouseScalaExample/src/main/scala/com/huawei/bigdata/spark/examples/SparkOnClickHouseExample.scala
package com.huawei.bigdata.spark.examples

import java.util.Properties

import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession

import com.clickhouse.jdbc.{ClickHouseDriver, ClickHouseStatement}

/**
 * SparkOnClickHouseScalaExample
 */

object SparkOnClickHouseExample {

  val DRIVER = "com.clickhouse.jdbc.ClickHouseDriver"

  def main(args: Array[String]): Unit = {
    if (args.length < 5) {
      println("Missing parameters, need ckJdbcUrl, ckDBName, ckTableName, userName, password")
      System.exit(0)
    }

    val spark =  SparkSession
      .builder()
      .config(getSparkConf)
      .appName("SparkOnClickHouseScalaExample")
      .enableHiveSupport()
      .getOrCreate()

    val url = args(0)
    val clickHouseDB = args(1)
    val clickHouseTable = args(2)
    val userName = args(3)
    val password = args(4)


    val props = new Properties()
    props.put("ssl", "true")
    props.put("user", userName)
    props.put("password", password)
    props.put("driver", DRIVER)
    props.put("isCheckConnection", "true")
    props.put("sslMode", "none")
    val ckDriver = new ClickHouseDriver
    val ckConnect = ckDriver.connect(url, props)
    val ckStatement = ckConnect.createStatement()

    val allDBs = s"show databases"
    val allDatabases = ckStatement.executeQuery(allDBs)
    while (allDatabases.next()) {
      val dbName = allDatabases.getString(1)
      println("dbName: " + dbName)
    }

    val startTime = System.currentTimeMillis()
    clickHouseExecute(ckStatement, clickHouseDB, clickHouseTable)
    println("[Elapsed]" + (System.currentTimeMillis() - startTime))

    val map = new java.util.HashMap[String, String]
    map.put("ssl", "true")
    map.put("user", userName)
    map.put("password", password)
    map.put("driver", DRIVER)
    map.put("isCheckConnection", "true")
    map.put("sslMode", "none")

    val ckData = spark.read
      .format("jdbc")
      .option("url", url)
      .options(map)
      .option("driver", DRIVER)
      .option("dbtable", clickHouseDB + "." + clickHouseTable)
      .load()

    ckData.show()

    ckData.registerTempTable("ckTempTable")

    val newCkData = spark.sql("select EventDate,cast(id+20 as decimal(20,0)) as id,name,age,address from ckTempTable")

    newCkData.show()
    newCkData.write.mode("Append").jdbc(url, clickHouseDB + "." + clickHouseTable, props)

    spark.stop()
  }

  def getSparkConf: SparkConf = {
    val conf = new SparkConf().setAppName("SparkOnClickHouseScalaExample")
    conf.set("spark.driver.maxResultSize", "1G")
    conf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    conf.set("spark.task.maxFailures", "20")
    conf.set("spark.default.parallelism", "50")
    conf.set("spark.executor.momory", "1G")
    conf.set("spark.driver.memory", "1G")
    conf.set("spark.executor.cores", "2")
    conf.set("spark.sql.shuffle.partitions", "50")
    conf.set("spark.hadoop.validateOutputSpecs", "false")
    conf
  }

  def executeSqlText(ckStatement: ClickHouseStatement, sqlText: String) {
    ckStatement.execute(sqlText)
  }

  def clickHouseExecute(ckStatement: ClickHouseStatement, clickHouseDB: String, clickHouseTable: String) {
    val createDB = s"CREATE DATABASE ${clickHouseDB} ON CLUSTER default_cluster"

    val createTable = s"CREATE TABLE ${clickHouseDB}.${clickHouseTable} ON CLUSTER default_cluster " +
      "(`EventDate` DateTime,`id` UInt64,`name` String,`age` UInt8,`address` String)" +
      s"ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/${clickHouseDB}/${clickHouseTable}', '{replica}')" +
      "PARTITION BY toYYYYMM(EventDate) ORDER BY id"

    val insertData = s"insert into ${clickHouseDB}.${clickHouseTable} (id, name,age,address) values (1, 'zhangsan',17,'xian'), (2, 'lisi',36,'beijing')"

    executeSqlText(ckStatement, createDB)
    executeSqlText(ckStatement, createTable)
    executeSqlText(ckStatement, insertData)
  }
}
